core/payments/reservation/reservation_ledger.go

```diff
@@ -0,0 +1,117 @@
+package reservation
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/Layr-Labs/eigenda/core"
+)
+
+// Tracks usage of a single account reservation
+//
+// This struct is goroutine safe.
+type ReservationLedger struct {
+	config ReservationLedgerConfig
+
+	// synchronizes access to the underlying leaky bucket algorithm
+	lock sync.Mutex
+
+	// an instance of the algorithm which tracks reservation usage
+	leakyBucket *LeakyBucket
+}
+
+// Creates a new reservation ledger, which represents the reservation of a single user with a leaky bucket
+func NewReservationLedger(
+	config ReservationLedgerConfig,
+	// now should be from a source that includes monotonic timestamp for best results
+	now time.Time,
+) (*ReservationLedger, error) {
+	leakyBucket, err := NewLeakyBucket(
+		config.reservation.symbolsPerSecond,
+		config.bucketCapacityDuration,
+		config.startFull,
+		config.overfillBehavior,
+		now,
+	)
+	if err != nil {
+		return nil, fmt.Errorf("new leaky bucket: %w", err)
+	}
+
+	return &ReservationLedger{
+		config:      config,
+		leakyBucket: leakyBucket,
+	}, nil
+}
+
+// Debit the reservation with a number of symbols.
+//
+// CheckInvariants should be called prior to calling Debit, to make sure the dispersal is permitted under the
+// parameters of the reservation. If CheckInvariants succeeds, then Debit is called to make sure the dispersal doesn't
+// exceed reservation capacity.
+//
+// Returns (true, nil) if the reservation has enough capacity to perform the debit.
+// Returns (false, nil) if the bucket lacks capacity to permit the fill.
+// Returns (false, error) if an error occurs. Possible errors include:
+//   - ErrTimeMovedBackward: current time is before a previously observed time (only possible if input time instances
+//     don't included monotonic timestamps)
+//   - Generic errors for all other unexpected behavior
+//
+// If the bucket doesn't have enough capacity to accommodate the fill, symbolCount IS NOT added to the bucket, i.e. a
+// failed debit doesn't count against the meter.
+func (rl *ReservationLedger) Debit(
+	// now should be from a source that includes monotonic timestamp for best results.
+	// This a local time from the perspective of the entity that owns this ledger instance, to be used with the local
+	// leaky bucket: it should NOT be sourced from the PaymentHeader
+	now time.Time,
+	// the number of symbols to debit
+	symbolCount uint32,
+) (bool, error) {
+	rl.lock.Lock()
+	defer rl.lock.Unlock()
+
+	success, err := rl.leakyBucket.Fill(now, symbolCount)
+	if err != nil {
+		return false, fmt.Errorf("fill: %w", err)
+	}
+
+	return success, nil
+}
+
+// Credit the reservation with a number of symbols. This method "undoes" a previous debit, following a failed dispersal.
+func (rl *ReservationLedger) RevertDebit(now time.Time, symbolCount uint32) error {
```

Comment:
Perhaps note that it might not perfectly undo things if time has passed, since a bucket is never allowed to go negative. So a revert does not necessarily mean you will actually be able to utilize all of the bandwidth you just reverted.

Reply:
Tried to clarify here df2cb309fdf1d118f3c6c661b77e0fc7ad2ce08c